From f31b3da495a98de373b02ade0a7fe1662dc9b492 Mon Sep 17 00:00:00 2001
From: kolun4ik <11_ka@rambler.ru>
Date: Sun, 9 Dec 2018 18:59:59 +0000
Subject: [PATCH 1/3] add opcode LOAD_OTUS into Include/opcode.h

---
 homework_02/Python-2.7.6/Include/opcode.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/homework_02/Python-2.7.6/Include/opcode.h b/homework_02/Python-2.7.6/Include/opcode.h
index 9764109..c3607f9 100644
--- a/homework_02/Python-2.7.6/Include/opcode.h
+++ b/homework_02/Python-2.7.6/Include/opcode.h
@@ -149,7 +149,7 @@ extern "C" {
 
 #define SET_ADD         146
 #define MAP_ADD         147
-
+#define LOAD_OTUS       148
 
 enum cmp_op {PyCmp_LT=Py_LT, PyCmp_LE=Py_LE, PyCmp_EQ=Py_EQ, PyCmp_NE=Py_NE, PyCmp_GT=Py_GT, PyCmp_GE=Py_GE,
 	     PyCmp_IN, PyCmp_NOT_IN, PyCmp_IS, PyCmp_IS_NOT, PyCmp_EXC_MATCH, PyCmp_BAD};
-- 
1.8.3.1


From 0b6758c2ca5878a3a143ad10338fbe964ed1ec2e Mon Sep 17 00:00:00 2001
From: kolun4ik <11_ka@rambler.ru>
Date: Sun, 9 Dec 2018 19:15:36 +0000
Subject: [PATCH 2/3] modify Python/peephole.c for replace LOAD_FAST 0
 LOAD_CONST n -> LOAD_OTUS

---
 homework_02/Python-2.7.6/Python/peephole.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/homework_02/Python-2.7.6/Python/peephole.c b/homework_02/Python-2.7.6/Python/peephole.c
index fb6cd03..2f0fd84 100644
--- a/homework_02/Python-2.7.6/Python/peephole.c
+++ b/homework_02/Python-2.7.6/Python/peephole.c
@@ -409,6 +409,14 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
                 cumlc = lastlc + 1;
                 break;
 
+	    case LOAD_FAST:
+		j = GETARG(codestr, i);
+		if (codestr[i+3] == LOAD_CONST && j == 0) {
+		    codestr[i+3] = LOAD_OTUS;
+		    memset(codestr+i, NOP, 3);
+		}
+		break;
+
                 /* Skip over LOAD_CONST trueconst
                    POP_JUMP_IF_FALSE xx. This improves
                    "while 1" performance. */
-- 
1.8.3.1


From 8d482ad9694101f3a87249b51ab7c082a22e9dbb Mon Sep 17 00:00:00 2001
From: kolun4ik <11_ka@rambler.ru>
Date: Sun, 9 Dec 2018 19:32:36 +0000
Subject: [PATCH 3/3] modify Python/ceval.c, add case LOAD_OTUS:....

---
 homework_02/Python-2.7.6/Python/ceval.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/homework_02/Python-2.7.6/Python/ceval.c b/homework_02/Python-2.7.6/Python/ceval.c
index 06ada97..cd4d1e4 100644
--- a/homework_02/Python-2.7.6/Python/ceval.c
+++ b/homework_02/Python-2.7.6/Python/ceval.c
@@ -1126,6 +1126,21 @@ PyEval_EvalFrameEx(PyFrameObject *f, int throwflag)
             PUSH(x);
             goto fast_next_opcode;
 
+	case LOAD_OTUS:
+	    x = GETLOCAL(0);
+	    if (x != NULL) {
+		Py_INCREF(x);
+		PUSH(x);
+		x = GETITEM(consts, oparg);
+		Py_INCREF(x);
+		PUSH(x);
+		goto fast_next_opcode;
+	    }
+	    format_exc_check_arg(PyExc_UnboundLocalError,
+		UNBOUNDLOCAL_ERROR_MSG,
+		PyTuple_GetItem(co->co_varnames, oparg));
+	break;
+
         PREDICTED_WITH_ARG(STORE_FAST);
         case STORE_FAST:
             v = POP();
-- 
1.8.3.1

